jdbi/src/main/java/io/micronaut/configuration/jdbi/spring/SpringTransactionHandler.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2020 original authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.micronaut.configuration.jdbi.spring;
+
+import io.micronaut.context.annotation.EachBean;
+import io.micronaut.context.annotation.Requires;
+import org.jdbi.v3.core.Handle;
+import org.jdbi.v3.core.HandleCallback;
+import org.jdbi.v3.core.transaction.TransactionHandler;
+import org.jdbi.v3.core.transaction.TransactionIsolationLevel;
+import org.springframework.transaction.PlatformTransactionManager;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.TransactionStatus;
+import org.springframework.transaction.support.DefaultTransactionDefinition;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Consumer;
+
+/**
+ * Allows Spring Transaction to be used with Jdbi.
+ *
+ * @author Dan Maas
+ * @since 1.3.1
+ */
+@Requires(classes = PlatformTransactionManager.class)
+@EachBean(PlatformTransactionManager.class)
```

Comment:
The Spring support is nice. I wonder would it make sense to add a dependency on `micronaut-data-tx` and include support for Micronaut's transaction management and Micronaut Data.

See https://search.maven.org/artifact/io.micronaut.data/micronaut-data-tx/1.0.0/jar

And https://github.com/micronaut-projects/micronaut-data/blob/master/data-tx/src/main/java/io/micronaut/transaction/jdbc/DataSourceTransactionManager.java

Reply:
@graemerocher would that be `instead` of using spring? It looks like micronaut's txn mgmt is mutually exclusive. If so, it might make sense to update the other libraries (like jooq) to use the same.

Reply:
I am also not sure how to test both of these in the same project due to the classpath exclusions. Would there be separate gradle modules for `jdbi-spring` and `jdbi-micronaut-data`?

Reply:
I can assume that the same would go for the jooq library

Reply:
Yeah you make some good points. We need to update JOOQ too

I think it is probably possible to do it in a single project without splitting if requirements are added like `@Requires(property="jdbi.spring.transactions", value="true", defaultValue="true")` and then you could set `jdbi.spring.transactions` to disable spring transaction management for the scope of the tests

Reply:
I'm still not sure how the Micronaut data transaction manager can be tested - it won't load if the spring transaction manager is on the classpath.

I am testing the micronaut-data transaction manager locally by toggling the gradle dependency, and for some reason rollback is not working. I can see the connection rollback happen, but the statements in the transaction get applied to the database anyways.

Reply:
I will have a look and get back to you

Reply:
So its cause by the lack of a transaction aware datasource proxy for the Micronaut impl. We have implemented this for the next version of Micronaut Data 1.0.1 

https://github.com/micronaut-projects/micronaut-data/pull/402

I guess to get it working for 1.0.0 of Micronaut Data you could wrap the datasource (if it is not an instance of `DelegatingDataSource` the same way as https://github.com/micronaut-projects/micronaut-data/pull/402/files#diff-ec1dcaa43f69ca2d01475a1d88349df5R24 does. Otherwise we can put off this feature until 1.0.1 is out and support it then.

Reply:
@graemerocher I think waiting until 1.0.1 is easier, and in the same feature jooq and anything else can also be upgraded to support micronaut data transactions.